### ISSUE TITLE
Ionic does not appear to support multiple views fully

### ISSUE THREAD

**Issue by skrpc** [driftyco/ionic#4925]
My ionic app uses both tabs and side menus, the side menus are dynamic and the content could be anything so I use multiple views, one view per tab and one view per side menu.

```
<ion-side-menus>
  <ion-side-menu-content>
    <ion-nav-bar class="bar-positive">
      <ion-nav-back-button ng-click="menu.changePage()" class=""></ion-nav-back-button>
    </ion-nav-bar>

    <ion-tabs>
      <ion-tab ng-click="menu.changePage('app.dashboard')" title="Dashboard" icon="ion-speedometer">
        <ion-nav-view name="dashboardContent"></ion-nav-view>
      </ion-tab>
      <ion-tab ng-click="menu.changePage('app.view_guides')" title="Guides" icon="ion-clipboard">
        <ion-nav-view name="guideContent"></ion-nav-view>
      </ion-tab>
      <ion-tab ng-click="menu.changePage('app.map')" title="Map" icon="ion-map">
        <ion-nav-view name="mapContent"></ion-nav-view>
      </ion-tab>
      <ion-tab ng-click="menu.changePage('app.more')" title="More" icon="ion-more">
        <ion-nav-view name="moreContent"></ion-nav-view>
      </ion-tab>
    </ion-tabs>
  </ion-side-menu-content>
  
  <ion-side-menu side="left" is-enabled="menu.leftMenuEnabled">
    <ion-nav-view name="leftMenuContent"></ion-nav-view>
  </ion-side-menu>
  
  <ion-side-menu side="right" is-enabled="menu.rightMenuEnabled">
    <ion-nav-view name="rightMenuContent"></ion-nav-view>
  </ion-side-menu>
</ion-side-menus>
```

In my router I can then target templates and controllers to each view

```
.state('app.show_guide', {
      url: '/show_guide',
      views: {
        'guideContent': {
          templateUrl: 'templates/guides/show-guide.html',
          controller: 'vrCtrl_fieldGuideShow as fieldGuideShow'
        },
        'rightMenuContent': {
          templateUrl: 'templates/page-menu.html',
          controller: 'vrCtrl_fieldGuideMenu as fieldGuideMenu'
        },
      }
    })
```

The first issue is I need a back button on the page but there are now two histories created, one for each view but there does not seem to be a way to define which history is the primary and therefore what the back should relate to.  What is missing is some way of indicating which is the primary view and therefore the one that $ionicHistory refers to.

Secondly I need to add an event handler to the $ionicView.enter and $ionicView.leave events but ionic does not call these events for both the views, it does call the events but only once on one of the views rather than once on both the views.

**Comment by ryannutley** [comment]
Just want to 1 up this request. Very similar issue. I have a sidebar menu that should only appear on one tab view, but there seems to be confusion on where the back link is trying to go to.

```javascript
.state('tab.item-detail', {
		url: '/items/:id',
		views: {
			'tab-items': {
				templateUrl: 'templates/item-detail.html',
				controller: 'ItemDetailCtrl'
			},
			'rightMenuContent': {
				templateUrl: 'templates/menu-chat.html',
				controller: 'ItemChatDetailCtrl'
			}
		}
	})
```

**Comment by chbomtempo** [comment]
I'm having the same issue. Anyone found a workarround for that limitation? Thanks in advance.

**Comment by danbucholtz** [comment]
Hi @skrpc,

We have made significant improvements to the way life-cycle events work in V1.  I am not sure we'll be able to get the history stuff in V1 as it would be a major change.

Can you test on the latest build?  Let me know if you need some help.

http://code.ionicframework.com/#nightly

I guess our staging site hasn't deployed in the past week or so, but we do have new and improved docs.
http://ionic-site-staging.herokuapp.com/docs/nightly/api/directive/ionView/

Basically, in addition to dispatching the event upwards (emit) from the correct scope, but we also now fire events downwards (broadcast) to work with ui-router nested views.

```
$scope.$on("$ionicParentView.enter", function(event, data){
   // woot, I received the event!
});
```

Please let me know if you have any issues.  I'll re-open if needed.

Thanks,
Dan

**Comment by skrpc** [comment]
I downloaded ionic 1.3.1 to try this and there appears to be another issue first.

My initial page is accessed from tab index 2 so I want that tab to be the default selected highlighted tab, I also access pages dynamically and need to make sure the correct tab is highlighted so I added the generic handler in a service

    $rootScope.$on(
      '$ionicView.enter',
      function(event, viewConfig) {
        if (
          $state.current.data && 
          typeof $state.current.data.tabIndex !== 'undefined' && 
          $ionicTabsDelegate.selectedIndex() != $state.current.data.tabIndex)
        {
          $ionicHistory.clearCache();
          $ionicTabsDelegate.select($state.current.data.tabIndex);
        }
      }
    );

The $ionicView.enter event handler now appears to not be fired for the very first page (i.e. on app load), I have tried using $ionicParentView.enter and this also is not fired for the first page on app load.  These events are fired for subsequent pages.

This means tab index 0 is highlighted and therefore the active view is the view for the first tab however the angular router and therefore url refer to a page on tab 2 which uses a different view so a blank page is shown.

**Comment by danbucholtz** [comment]
@skrpc, can you make a codepen showing the issue?  In 1.3.1, we made significant changes to the way the life cycle events are dispatched.  We wrote several thousand lines of automated tests to verify the changes, so we must have missed a test case.  [See changes here](https://github.com/driftyco/ionic/commit/d63733350b73c0084f99fe285dda96f0204b2cd7#diff-82d6d3da2a7bb1d17947d4e25aa93923).

Thanks,
Dan

**Comment by skrpc** [comment]
[http://codepen.io/skrpc/pen/MyMWRv]http://codepen.io/skrpc/pen/MyMWRv

It looks a bit messy but basically I had a whole world of issues with ionic when using views that required targeting absolute views (i.e. ???@app) as my full app also uses side menus and this is where I got to so my code would work.

I found I have to navigate by always going through my own changePage() function and no longer using href='' but instead always using an ng-click on buttons and links.

Anyway, when using ionic 1.3.0 $rootScope.$on('$ionicView.enter', ...) is called when the app first starts and the Second tab and view are made the active ones (as a result of the use of $urlRouterProvider.otherwise('/tabSecond');).

When using ionic 1.3.1 it is not called and the only way to get the second tab and view to be the active ones is to uncomment the $timeout in the run function but this is not an answer as then I am guessing on the delay whilst also adding a delay in the first place.
